code/HabitTracker/app/src/main/java/com/example/habittracker/DatabaseManager.java

```diff
@@ -240,4 +251,162 @@ public void onFailure(@NonNull Exception e) {
+                if (task.isSuccessful()) {
+                    ArrayList<HabitEvent> eventArray = new ArrayList<>();
+                    for (QueryDocumentSnapshot doc : task.getResult()) {
+                        Date date;
+                        try {
+                             date = new SimpleDateFormat("yyyy-MM-dd").parse((String)doc.getData().get("startDate"));
+                        } catch (Exception e) {
+                            date = new Date();
+                        }
+                        eventArray.add(new HabitEvent(
+                                habit,
+                                (String)doc.getData().get("comment"),
+                                date,
+                                (String)doc.getData().get("comment"),
+                                "image"
+                        ));
+                    }
+                    callback.onCallbackSuccess(eventArray);
+                }
+                else{
+                    callback.onCallbackFailed();
+                }
+            }
+        });
+    }
+
+    public void getUserDetails(String username, UserDetailsCallback callback){
+        // Users -> userid (key) -> Habits -> habitTitle (key) -> HabitEvents
+        DocumentReference doc = usersColRef
+                .document(username);
+        doc.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                if (task.isSuccessful()) {
+                    User user;
+                    DocumentSnapshot document = task.getResult();
+                    if(!document.exists()){
+                        callback.onCallbackFailed();
+                        return;
+                    }
+                    String hashedPassword = (String) document.getData().get("hashedPassword");
+                    user = new User(
+                            username,
+                            hashedPassword
+                    );
+                    callback.onCallbackSuccess(user);
+                }
+                else{
+                    callback.onCallbackFailed();
+                }
+            }
+        });
+    }
+
+    public void userExists(String username, BooleanCallback callback){
+        // Users -> userid (key) -> Habits -> habitTitle (key) -> HabitEvents
+        DocumentReference doc = usersColRef
+                .document(username);
+        doc.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                if (task.isSuccessful()) {
+                    boolean flag = false;
+                    DocumentSnapshot document = task.getResult();
+                    if(document.exists()){
+                        flag = true;
+                    }
+                    callback.onCallbackSuccess(flag);
```

Comment:
I was hoping to see a UserDetailsCallback with methods onCallbackSuccess(String username, String hashedPassword). 
This way, the controller class like the LoginActivity can create the User object inside its implementation of UserDetailsCallback.onCallbackSuccess and start a new Activity from within the callback.
It might be harder for the controller class to take actions inside its implementation of UserDetailsCallback.onCallbackSuccess while accessing variables that lie outside of the callback function. If I recall correctly, if you want use an outer class variable inside an anonymous inner class (which is the callback implementation), the outer class variables need to be declared final.

Reply:
I gotcha, I will make a new branch and make the requested changes. since this is after the merge.

Reply:
@NickSNFK Did you make the changes?

Reply:
Yes but I’m not at home right now. I will be in like an hour to work on this.